### PR TITLE
feat: Rolling window splitter

### PR DESCRIPTION
**Pull request by simjak**
A splitter that divides documents into segments based on semantic similarity
using a rolling window approach.
It adjusts the similarity threshold dynamically.
Splitting is based:
- On the similarity threshold
- On the maximum token limit for a split

Notebook:
`semantic-router/docs/examples/text_semantic_split.ipynb`

```
from semantic_router.splitters import RollingWindowSplitter
from semantic_router.encoders import OpenAIEncoder

splitter = RollingWindowSplitter(
    encoder=OpenAIEncoder(),
    min_split_tokens=50,
    max_split_tokens=300,
    window_size=5, # sentences
    plot_splits=True)
    
splits = splitter([text])
```

If you want to print splitted texts
```
splitter.print(splits)
```

image

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -51,6 +53,8 @@ pytest-xdist = "^3.5.0"
 mypy = "^1.7.1"
 types-pyyaml = "^6.0.12.12"
 types-requests = "^2.31.0"
+termcolor = "^2.4.0"
```
Comment (jamescalam) [inline pyproject.toml:56]: Can we only use colorama? I think both do the same thing

File: semantic_router/index/pinecone.py [hunk 2]
```diff
@@ -50,14 +50,10 @@ class PineconeIndex(BaseIndex):
...
-        self.type = "pinecone"
-        self.client = self._initialize_client()
-        if not self.index_name.startswith(self.index_prefix):
-            self.index_name = f"{self.index_prefix}{self.index_name}"
-        # Create or connect to an existing Pinecone index
-        self.index = self._init_index()
+        clean_route = clean_route_name(self.route)
+        # Use SHA-256 for a more secure hash
+        utterance_id = hashlib.sha256(self.utterance.encode()).hexdigest()
+        self.id = f"{clean_route}#{utterance_id}"
```
Comment (jamescalam) [inline semantic_router/index/pinecone.py:56]: This seems wrong? We are no longer initialising the index and are creating a self.id attribute?

File: semantic_router/splitters/rolling_window.py [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+                logger.info(f"Iteration {iteration}: Adjusting high to {high}")
+            else:
+                low = self.calculated_threshold + self.threshold_adjustment
+                logger.info(f"Iteration {iteration}: Adjusting low to {low}")
+            iteration += 1
+
+        logger.info(f"Final optimal threshold: {self.calculated_threshold}")
+        return self.calculated_threshold
+
+    def split_documents(
```
Comment (jamescalam) [inline semantic_router/splitters/rolling_window.py:118]: Should be __call__ method to align with other splitters

File: semantic_router/splitters/utils.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        # OR
+        |
+        # Handles splitting after ellipses
+        (?<=\.\.\.)\s+(?=[A-Z])
+        # OR
+        |
+        # Matches and removes control characters and format characters
+        [\p{Cc}\p{Cf}]+
+    """
+    sentences = regex.split(regex_pattern, text, flags=regex.VERBOSE)
```
Comment (jamescalam) [inline semantic_router/splitters/utils.py:55]: Very cool

**Review by jamescalam — CHANGES_REQUESTED** [review]
Some minor tweaks needed, otherwise looks great!

File: semantic_router/splitters/rolling_window.py (hunk 3 above)
Comment (simjak) [reply semantic_router/splitters/rolling_window.py:118]: no, I have __call__ method:
```
    def __call__(self, docs: List[str]) -> List[DocumentSplit]:
        if not docs:
            raise ValueError("At least one document is required for splitting.")

        if len(docs) == 1:
            token_count = tiktoken_length(docs[0])
            if token_count > self.max_split_tokens:
                logger.warning(
                    f"Single document exceeds the maximum token limit "
                    f"of {self.max_split_tokens}. "
                    "Splitting to sentences before semantically splitting."
                )
            docs = split_to_sentences(docs[0])
        encoded_docs = self.encode_documents(docs)
        self.find_optimal_threshold(docs, encoded_docs)
        similarities = self.calculate_similarity_scores(encoded_docs)
        split_indices = self.find_split_indices(similarities=similarities)
        splits = self.split_documents(docs, split_indices, similarities)
        self.plot_similarity_scores(similarities, split_indices)
        return splits
```

**Review by jamescalam — APPROVED** [review]

**Pull request merged**